feat: js-peer-workshop

Nkovaturient opened the pull request:
## ✨ Add Complete JS-peer Universal Connectivity Workshop

## **Description**

This PR introduces the complete **Js-peer Universal Connectivity Workshop** (Lessons 1-8). These lessons provide a comprehensive, hands-on path for developers to build robust, event-driven peer-to-peer applications using [js-libp2p](https://github.com/libp2p/js-libp2p). The workshop guides students from basic node creation to a fully functional decentralized chat application, mirroring the high standards and structure of the Rust workshop.

---

## **Integration Highlights**

- **Lesson 1: Identity & Swarm Basics**
  - **Goal**: Initialize a libp2p node and generate cryptographic identity.
  - **Features**: PeerId generation, basic swarm setup, multiaddr listening.

- **Lesson 2: TCP Transport**
  - **Goal**: Establish reliable connections between peers.
  - **Features**: TCP transport configuration, Noise encryption, Yamux stream multiplexing.

- **Lesson 3: Ping Protocol**
  - **Goal**: Monitor connection health and latency.
  - **Features**: Integration of the Ping service, latency measurement, connection liveness checks.

- **Lesson 4: Circuit Relay v2**
  - **Goal**: Enable connectivity across NATs and firewalls.
  - **Features**: configuring Relay nodes, Listeners (reservations), and Dialers for hole punching.

- **Lesson 5: Identify Protocol**
  - **Goal**: Dynamic capability discovery and protocol negotiation.
  - **Features**: Automatic peer information exchange (user agent, protocol versions, public keys).

- **Lesson 6: Gossipsub Module**
  - **Goal**: Efficient scalable message broadcasting.
  - **Features**: PubSub configuration, topic subscription, mesh network formation for chat.

- **Lesson 7: Kademlia DHT**
  - **Goal**: Decentralized peer discovery and content routing.
  - **Features**: Kademlia DHT setup, peer routing (finding peers by ID), content routing (providing/finding CIDs).

- **Lesson 8: Final Checkpoint (Chat Application)**
  - **Goal**: Build a production-ready P2P chat app.
  - **Features**: Full stack integration (TCP+Noise+Yamux+Ping+Identify+Gossipsub+DHT). Includes an interactive **Live Checker** server for real-time validation and chat.

- **Automation & Testing**
  - **Unified Checker**: Uses a Docker-based checking system (`check.py` + `docker-compose`) for all lessons.
  - **Validation**: Automated scripts verify output logs, peer IDs, and connectivity states.
  - **Reproducibility**: `deps.py` and `setup.md` ensure consistent developer environments.

- **Structure & Consistency**
  - Follows the standard `en/js/0X-...` directory structure.
  - Each lesson contains a dedicated `lesson.md`, `app/` code skeleton, and validation scripts.

---

## **Relevant Resources**

- [js-libp2p Getting Started Guide](https://docs.libp2p.io/guides/getting-started/javascript)
- [js-libp2p API Docs](https://libp2p.github.io/js-libp2p/)
- [Universal Connectivity JS Peer Reference](https://github.com/libp2p/universal-connectivity/tree/main/js-peer)

Nkovaturient committed:
feat(js-peer): integrate-identity-and-tcp-lessons

Nkovaturient committed:
Update scripts.md

dhuseby reviewed (changes requested):
Thank you for this PR! However, it is missing the `en/js/setup.md` file that explains the commands to run to start a new project that uses js-libp2p and the `en/js/deps.py` script that checks for all of the commands needed to create/compile/run a js project that uses js-libp2p. I can't test it any further until those are there.

Nkovaturient commented:
Thank you for the review, working on it!

Nkovaturient committed:
feat(js-workshop): add lessons 1–3, deps.py + setup.md

Nkovaturient committed:
resolve-merge-conflicts

Nkovaturient commented:
Hello @dhuseby 

## QnA

1) Need your help in correctly configuring the `setup.md`, ig for each workshop, the dir `universal-connectivity` is created and the subsequent `rust/js/py` lessons are setup, right? Can you look at the current setup.md for `js-lessons` and provide guidance on it.

2) How does `Checkpoint lessons` should be graded/implemented?scores? need clarification.


### As for the latest commits inculcates following changes: 

- Scaffolded and implemented Lesson 1 (Identity & Swarm Basics) with Docker, npm, and robust validation
- Scaffolded and implemented Lesson 2 (TCP Transport) with error handling, event-driven logic, and stack consistency
- Scaffolded and implemented Lesson 3 (Ping Protocol, Checkpoint 1) with identify, ping, and direct-message protocols
- Added direct-message.js: production-grade, event-driven, error-resilient direct message protocol implementation

### **Further TODOs**

- Scaffold and implement future lessons (Gossipsub, DHT, NAT traversal, etc.).
- Add instructor checkpoint integration for live classroom validation.
- Expand automated CI coverage for all lessons.


### Console Output:

## Lesson 1: (Identity & Swarm Basics)
 
![Screenshot (611)](https://github.com/user-attachments/assets/860cb547-68bc-47bb-ba42-72165abae35f)

## Lesson 2 (TCP Transport)

![Screenshot (610)](https://github.com/user-attachments/assets/c4383a9d-a16a-477f-8e5d-1f6718f752a4)

## Lesson 3 (Ping Protocol)

![Screenshot (609)](https://github.com/user-attachments/assets/575d3e5d-0c3e-46e1-acc7-ba392d797e3c)

luu-alex committed:
add to lesson

Nkovaturient committed:
Merge pull request #1 from luu-alex/update/js-peer-workshop

revamp lesson.md in a comprehensive way.

luu-alex commented:
@dhuseby Hey there! I'll help create the set up/checks and deps for js

luu-alex committed:
update docker files

luu-alex committed:
update setup

luu-alex committed:
Merge branch 'feat/js-peer-workshop' into update/js-peer-workshop

luu-alex commented:
Created a PR to edit the docker files and configs to structure for the workshop flow.
https://github.com/Nkovaturient/universal-connectivity-workshop/pull/2

Nkovaturient committed:
Merge pull request #2 from luu-alex/update/js-peer-workshop

Update/js peer workshop

Nkovaturient committed:
Merge branch 'main' of https://github.com/Nkovaturient/universal-connectivity-workshop into feat/js-peer-workshop

luu-alex committed:
update check and docker files

luu-alex committed:
Merge pull request #3 from luu-alex/update/js-peer-workshop

Update lesson 2 dockerfiles

Nkovaturient committed:
feat(js-peer):setup lesson-4-to-6 + modify check.py

Nkovaturient committed:
Merge branch 'feat/js-peer-workshop' of https://github.com/Nkovaturient/universal-connectivity-workshop into feat/js-peer-workshop

Nkovaturient commented:
## [Updates](https://github.com/libp2p/universal-connectivity-workshop/pull/2/commits/f625cd979ef36d5db0a7e54339de1c76a35a48d2)

### 03-ping-protocol/check.py
- **Purpose:** Validates the implementation of the ping protocol.
  - Provides clear, actionable feedback for each step and prints a summary of what was achieved.

---

### 04-circuit-relay-v2
- Demonstrates p2p connections between two peers where one(listener) peer is behind a NAT/firewall(unreachable) however can be connected via a relay reservations medium tht sends inbound connections from listener node to dialer node, facilitating communications. 

---

### 05-identify-protocol
- Implements identify protocol lesson tht facilitates exchange of  info with other peers in the network, including addresses, public keys, and capabilities.

---

### 06-gossipsub-module
- Lesson 6 implements extensible baseline pubsub protocol: `gossipsub` to clarify how p2p connections and messages are communicated between connected peers in a network.

---

### General Setup/Pattern
- Each lesson folder contains:
  - A robust `check.py` script tailored to the protocol/module being taught.
  - An `app/` directory with the actual node implementations for the lesson.

### Core features are setup just Lesson 4 and 5 are under active development  ..🚧

luu-alex committed:
update index.js

luu-alex committed:
Merge remote-tracking branch 'upstream/main' into update/js-peer-workshop

luu-alex committed:
Merge branch 'feat/js-peer-workshop' into update/js-peer-workshop

luu-alex committed:
update, add more information to  lesson and be compatible with docker

luu-alex committed:
Merge pull request #4 from luu-alex/update/js-peer-workshop

Update/js peer workshop lesson 2

luu-alex commented:
finished lesson 2, added more relevant info and fixed docker. added checker as well

luu-alex committed:
update lesson three

luu-alex committed:
Merge pull request #5 from luu-alex/update/js-peer-workshop

Update/js peer workshop lesson 3

luu-alex committed:
update lesson 1

luu-alex committed:
Merge pull request #6 from luu-alex/update/js-peer-workshop

update lesson 1

luu-alex committed:
update lesson 2

luu-alex committed:
Merge pull request #7 from luu-alex/update/js-peer-workshop

update lesson 2

dhuseby reviewed (changes requested):
Still need to clean some stuff up.

Nkovaturient committed:
Merge branch 'libp2p:main' into feat/js-peer-workshop

Nkovaturient committed:
chore: update js workshop setup and remove test scripts

Nkovaturient committed:
Merge branch 'feat/js-peer-workshop' of https://github.com/Nkovaturient/universal-connectivity-workshop into feat/js-peer-workshop

Nkovaturient commented on en/js/setup.md:
```diff
@@ -0,0 +1,23 @@
+| npm                   | ≥ 9                 |
+| TypeScript (optional) | ≥ 5.4               |
+
+
```
modified it,
 is it correct henceforth?

Nkovaturient commented:
Hey @dhuseby 
I have revamped the js-lessons setup
- Updated `deps.py` - Removed Docker and docker-compose checks, now only validates `Node.js and npm`
- Modified ` setup.md` - Added proper heading structure with `## 1. Project Setup` for the directory creation section
- Modified `lesson.yaml` - to include `status` indicators
- Cleaned up structure - Removed test scripts

Thanks for the [Workshop_Authoring_MD](https://github.com/libp2p/workshop/blob/main/WORKSHOP_AUTHORING.md),  will go through it.
 Is there anything else to take care of?

parth-soni07 committed:
Lesson-3 Fixes

parth-soni07 committed:
Minor fixes to console logs

Nkovaturient committed:
fix: lesson-3

Merge PR #8 from Nkovaturient/feat/lesson-3-fixes

Hany-Almnaem committed:
fix(l5): Identify-Protocol-checkpoint

* feat: Add Lesson 5 (Identify Protocol) for JavaScript track

- Complete lesson implementation with libp2p identify protocol
- Fixed PeerId generation and private key handling
- Added Docker containerization with proper networking
- Included comprehensive validation with check.py
- Updated workshop.yaml to support JavaScript lessons
- All tests passing: Node.js direct, Docker Compose, Python checker

Resolves: JavaScript Lesson 5 implementation
Testing: Validated with multiple test methods

* Add js-libp2p support and complete lesson paths for js in workshop.yaml

Nkovaturient committed:
fix: lesson-5 (#11)

* fix(l5): Identify-Protocol-checkpoint

* feat: Add Lesson 5 (Identify Protocol) for JavaScript track

- Complete lesson implementation with libp2p identify protocol
- Fixed PeerId generation and private key handling
- Added Docker containerization with proper networking
- Included comprehensive validation with check.py
- Updated workshop.yaml to support JavaScript lessons
- All tests passing: Node.js direct, Docker Compose, Python checker

Resolves: JavaScript Lesson 5 implementation
Testing: Validated with multiple test methods

* Add js-libp2p support and complete lesson paths for js in workshop.yaml

* Revert "fix(l5): Identify-Protocol-checkpoint"

This reverts commit 85ca2c33573f9931849cea719dc61c413170953b.

---------

Co-authored-by: Hany Almnaem <Almnaemhany333@gmail.com>

parth-soni07 committed:
chore(docker):L3+4 docker config and refactoring

* Added Lesson 03 & 04 with dockerization

* Updated peer-id generation logic

* minor changes to lesson03 app/index.js

Hany-Almnaem committed:
feat: add JavaScript Lesson 6 - GossipSub Pub/Sub

Hany-Almnaem committed:
Merge branch 'lesson-6-gossipsub' into addlesson6-on-workshop

Hany-Almnaem committed:
Merge branch 'feat/js-peer-workshop' of https://github.com/Nkovaturient/universal-connectivity-workshop into addlesson6-on-workshop

Hany-Almnaem committed:
Revert "Merge branch 'feat/js-peer-workshop' of https://github.com/Nkovaturient/universal-connectivity-workshop into addlesson6-on-workshop"

This reverts commit 6ab5765cf3bdfe9223d19adadd2a1e1f2fe468dd, reversing
changes made to 669e5a92ecd4327794a979d36276039ffbc78c0f.

Hany-Almnaem committed:
feat: add JavaScript Lesson 6 - GossipSub Pub/Sub (#14)

Hany-Almnaem committed:
feat: enhance GossipSub lesson with multi-peer demonstration (#15)

* feat: enhance GossipSub lesson with multi-peer demonstration

Transform basic GossipSub lesson into comprehensive multi-peer demo:

- Replace single-peer setup with realistic 2-peer network
- Demonstrate actual peer-to-peer message exchange
- Add professional educational documentation
- Implement proper mesh formation and validation
- Show bidirectional GossipSub messaging in action

Educational improvements:
- Students experience real decentralized messaging
- See authentic libp2p peer discovery and connection
- Understand GossipSub mesh topology formation
- Validate actual pub/sub message propagation

The lesson now demonstrates true GossipSub capabilities instead
of basic topic subscription, providing realistic P2P learning.

* feat: implement coordinated multi-peer GossipSub demonstration

Commit:
add lesson 07: kad-dht

Nkovaturient commented:
Hey @paschal533 ,
indeed grateful for your review and pointers ahead. 
on the same note, wish to update you that:-

 **Note 1:** QUIC support in js-libp2p is still experimental, hence after discussing with manusheel, we didnt include lesson for QUIC transport here.
**Note 2:**  Kad-dht checkpoint done
**Note 3:** Final Checkpoint lesson to be completed soon.

Commit:
add lesson-08:final checkpoint

Commit:
feat(lesson-08):fix gossipsub mesh and add interactive checker server

paschal533 commented:
> Hey @paschal533 thanks a lot for your detailed guidance and timely inputs. the issue is fixed + I have improvised the checker ln an interactive manner.
> 
> <img alt="Screenshot 2025-10-23 at 9 50 37 PM" width="1709" height="694" src="https://private-user-images.githubusercontent.com/127786136/504831218-4c0ef493-8ff1-477a-8b15-fcbb791eb843.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjEyMzc1NjEsIm5iZiI6MTc2MTIzNzI2MSwicGF0aCI6Ii8xMjc3ODYxMzYvNTA0ODMxMjE4LTRjMGVmNDkzLThmZjEtNDc3YS04YjE1LWZjYmI3OTFlYjg0My5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMDIzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTAyM1QxNjM0MjFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02YTk0ZjU0OTFiYWYyM2EzZjBkODFiOWE0Yjg4Yjc1Mjg5MTQ0NmNmZDM3MzI5Zjc3OTkyODAwYjI0MjhiNzdiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.ZjbCjjXcNq-S-zK3q8tUV_prNlIMgwGknhO99ruOTIQ">

Well done @Nkovaturient 👍

dhuseby commented:
@paschal533 @Nkovaturient I updated the rust version of the workshop so that there is a single checker application in the `en/checker` folder because it is in English. I restructured all of the Rust lessons so that they use the single checker application+docker image and each have their own `check.py` script for checking completion of the lesson.

I would like you to refactor this PR so that it uses the single checker docker image in the docker-compose.yaml files. You can look at the rust lessons for examples. It works as it used to, except that it has all of the protocols and outputs log messages in a standard format that can be used in `check.py` scripts across all lessons. This ensures that what the students end up with, is capable of connecting to the actual universal connectivity bootstrap peers and will run.

Let me know when it's rebased and refactored and I'll test and land it.

Nkovaturient committed:
Merge branch 'libp2p:main' into main

Nkovaturient committed:
Merge branch 'libp2p:main' into feat/js-peer-workshop

Commit:
fix merge conflicts

Commit:
feat(js-peer): refactor lessons, implement unified checker, interactive lesson-8-checker

Commit:
merge: complete merge with resolved conflicts

Nkovaturient commented:
hey @dhuseby kindly review
added `run-all-lessons.sh` and `setup.sh` temporarily for you to run and verify js-workshop lessons working.

> I would like you to refactor this PR so that it uses the single checker docker image in the docker-compose.yaml files. You can look at the rust lessons for examples. It works as it used to, except that it has all of the protocols and outputs log messages in a standard format that can be used in `check.py` scripts across all lessons. This ensures that what the students end up with, is capable of connecting to the actual universal connectivity bootstrap peers and will run.

hey @paschal533 ,
could you review and test if the js-workshop are correctly refactored and rebased as asked?

paschal533 commented:
HI @Nkovaturient, Great job on refactoring and rebasing the PR as requested. I have a few suggestions:

1. The `Checker` folder is not needed here anymore. @dhuseby created a single checker folder for all the lessons in English.
2. You don't also need to include the `rs` folder in this PR.
3. Run the universal connectivity commands on your system:

```
cargo install workshop

workshop --install https://github.com/libp2p/universal-connectivity-workshop
```
Then copy this Js folder and paste it inside the `en` folder. Then select `Javascript` as the lesson you want to take, go through the lessons and let me know if there is anything you need my help to fix.

4. You should also remove the lesson 8 for now. We will be adding it later in a better way.

Great job on the efforts you are putting in this PR. Your lessons are awesome!

Nkovaturient committed:
Merge branch 'libp2p:main' into feat/js-peer-workshop

paschal533 committed:
fix: Unauthorized issue and a few others

Nkovaturient committed:
chore: update setup.md, workshop.yaml, &  remove checker

Nkovaturient committed:
Merge paschal533/feat/js-peer-workshop
